src/main/java/net/javadiscord/javabot/util/IndentationHelper.java

```diff
@@ -0,0 +1,205 @@
+	 *
+	 * @param text The text that should be indented.
+	 * @param type The type of indentation to be used.
+	 * @return The indented String with the format specified.
+	 */
+	public static String formatIndentation(String text, IndentationType type) {
+		return new IndentationHelper(text, type).formatIndentation();
+	}
+
+	private String formatIndentation() {
+		if (type == IndentationType.NULL) {
+			return text;
+		}
+		IndentationState currentState = IndentationState.CODE;
+		for (int i = 0; i < text.length(); i++) {
+			char current = text.charAt(i);
+			if (startOfLine && current == ' ') {
+				continue;
+			}
+			builder.append(current);
+			if (current == '\n') {
+				builder.append(type.getPattern().repeat(Math.max(numberOfBrackets, 0)));
+				startOfLine = true;
+			}
+			currentState = switch (currentState) {
+				case CODE ->
+					processTokenInCode(i, current);
+				case STRING -> {
+					if (current == '\"' && !isEscaped(builder, builder.length() - 1)) {
+						yield IndentationState.CODE;
+					}
+					yield IndentationState.STRING;
+				}
+				case CHARACTER -> {
+					if (current == '\'' && !isEscaped(builder, builder.length() - 1)) {
+						yield IndentationState.CODE;
+					}
+					yield IndentationState.CHARACTER;
+				}
+				case SINGLE_LINE_COMMENT ->
+					current == '\n' ? IndentationState.CODE : IndentationState.SINGLE_LINE_COMMENT;
+				case MULTI_LINE_COMMENT -> {
+					if (current == '*' && i + 1 < text.length() && text.charAt(i + 1) == '/') {
+						yield IndentationState.CODE;
+					}
+					yield IndentationState.MULTI_LINE_COMMENT;
+				}
+			};
+			if (!Character.isWhitespace(current) && startOfLine) {
+				startOfLine = false;
+			}
+		}
+		return builder.toString();
+	}
+
+	private IndentationState processTokenInCode(int i, char current) {
+		switch (current) {
+			case '{' -> numberOfBrackets++;
+			case '}' -> {
+				numberOfBrackets--;
+				if (startOfLine && builder.length() - type.getNumberOfChars() - 1 >= 0) {
+					builder.replace(builder.length() - type.getNumberOfChars() - 1, builder.length(), "}");
+				}
+			}
+			case '\'' -> {
+				return IndentationState.CHARACTER;
+			}
+			case '\"' -> {
+				return IndentationState.STRING;
+			}
+			case '/' -> {
+				if (i + 1 < text.length()) {
+					if (text.charAt(i + 1) == '/') {
+						return IndentationState.SINGLE_LINE_COMMENT;
+					} else if (text.charAt(i + 1) == '*') {
+						return IndentationState.MULTI_LINE_COMMENT;
+					}
+				}
+			}
+		}
+		return IndentationState.CODE;
+	}
+
+	/**
+	 * Determines if the character in the StringBuilder at the specified position is escaped.
+	 *
+	 * @param builder the StringBuilder which holds the current String
+	 * @param index   The index at which the character to be checked is located.
+	 * @return True, if the escape character is referring to the character at the index, false otherwise.
+	 */
+	private static boolean isEscaped(StringBuilder builder, int index) {
```

Comment:
Remove `index` as it is always `builder.length() - 1`.
That would also remove the need for the `index >= builder.length()` check (which would never be executed).
If `index` would be initialized with `builder.length() - 2`, the `index--;` decrement would also not be necessary any more.
With these changes, the `while` loop can be changed to a standard for loop like
```java
for (int index=builder.length() - 2; index > 0 && builder.charAt(index) == '\\'; index--){
    numberOfCharacters++;
}
```

Reply:
I would leave it as is for now. Yes it could be simplified but I want to leave it as a utility method as it doesn't readability.

Reply:
Why do you think it doesn't improve readability to replace the whole method with something like
```java
private static boolean isEscaped(StringBuilder builder) {
	for (int index=builder.length() - 2; index > 0 && builder.charAt(index) == '\\'; index--){
    numberOfCharacters++;
	}
	return numberOfCharacters % 2 == 1;
}
```